openedx_tagging/core/tagging/rest_api/v1/serializers.py

```diff
@@ -184,6 +250,33 @@ def get_sub_tags_url(self, obj: TagData | Tag):
             return request.build_absolute_uri(url)
         return None
 
+    @property
+    def _model(self) -> Type:
+        """
+        Returns the model used when checking permissions.
+        """
+        return Tag
+
+    def get_can_change(self, instance) -> Optional[bool]:
+        """
+        Returns True if the current user is allowed to edit/change this Tag instance.
+
+        Returns False for all TagData instances.
+        """
+        if isinstance(instance, Tag):
+            return super()._can('change', instance)
+        return None
```

Comment:
from [comment](https://github.com/openedx/openedx-learning/pull/138#issuecomment-1879332200):

> The permissions for all tags in a taxonomy are always the same, right? ...  So I'm tempted to say don't include permissions for them - just include perms at the taxonomy level or course component level. But I realize that's making an assumption that this won't change in the future. If we include the perms now with every Taxonomy/Tag/ObjectTag as you're proposing, and make the frontend work accordingly, then we can be super flexible about how permissions work in the future if we want to, and won't need to make any changes to the frontend. So that makes sense.

Here is where that plan breaks down, because we don't currently have a way to check use permissions for TagData instances (rather than Tag instances). I could create a dummy Tag here from the TagData instance and use that to check the permissions, but that will likely trigger another database query (for the `Tag.taxonomy` field).

But we could use the UserPermissionsPagination class to put permissions for the taxonomy itself into the top level like we did for `can_add`?

Reply:
@pomegranited For now, the permissions of every tag in the taxonomy are the same. But we want to design the REST API to be flexible in the future, as I mentioned. So I would suggest this:

When you initialize a `TagDataSerializer`, pass in the `can_change` and `can_delete` permissions as `context`, just the same way that we currently pass in `taxonomy_id` and `request` as context. Then just change `get_can_change` and `get_can_delete` to return the value from the context, or omit those fields if the data is not available in the context. This will be simple, complete, and works with both `TagData` and `Tag`. It also maintains a flexible API so that if we need perms that vary per-tag in the future, we can change this to do that somehow, without needing to change the REST API or the frontend.

Reply:
I'm failing to find a hypothetical case where we need different Tag and Taxonomy permissions. Do you see any @pomegranited and @bradenmacdonald?

If we can't foresee a use for different permissions, I think we could move closer to removing the Tag permissions and keeping it simple instead of making it more flexible. Of course, only if we can't see a use for it in the future.

So, now, I'm more inclined to use the Taxonomy `canTagObject` permission for Tagging and the Taxonomy `canChange` for Tag editing (that we don't have frontend yet).

Reply:
@rpenido 

> So, now, I'm more inclined to use the Taxonomy canTagObject permission for Tagging and the Taxonomy canChange for Tag editing (that we don't have frontend yet), skipping serializing permission at TagDataSerializer.

I ended up going with @bradenmacdonald 's suggestion for the TagData:

> When you initialize a TagDataSerializer, pass in the can_change and can_delete permissions as context, just the same way that we currently pass in taxonomy_id and request as context. 

...so we have the permissions data in the REST API whenever we start editing taxonomy tags in the UI.

But I'm equally ok with removing this work if we feel like it's more unused code we shouldn't be maintaining yet.

Reply:
Hi, @pomegranited!
While reviewing, it took me some time to get the permission context again, and I felt we were overengineering a bit with some "to be used in the future" features.
 
If you (who wrote this change) didn't feel this way, we probably don't need to simplify it. If this is the case, I'm ok with the current solutions!

Reply:
Thank you for your thoughts here @rpenido ! I'll leave this comment thread unresolved so upstream/CCs can see it too, and will see what they say.